src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -142,6 +147,74 @@ public async Task StopContainersAsync(IExecutionContext executionContext, object
+            using HttpClient httpClient = new HttpClient(httpClientHandler);
+            httpClient.DefaultRequestHeaders.Add("Content-Type", "application/x-www-form-urlencoded");
+            using StringContent stringContent = new StringContent("grant_type=access_token&service=" + registryServer[8..] + "&tenant=" + tenantId + "&access_token=" + AADToken);
+            string ACRpassword = string.Empty;
+            int waitedTime = 0;
+            do
+            {
+                try
+                {
+                    var response = await httpClient.PostAsync(url, stringContent, cancellationToken).ConfigureAwait(false);
+                    if (response.StatusCode == HttpStatusCode.OK)
+                    {
+                        Dictionary<string, string> list = (Dictionary<string, string>)JsonConvert.DeserializeObject(await response.Content.ReadAsStringAsync(), typeof(Dictionary<string, string>));
+                        ACRpassword = list["refresh_token"];
+                    }
+                    else if (response.StatusCode == HttpStatusCode.TooManyRequests || response.StatusCode == HttpStatusCode.InternalServerError)
+                    {
+                        if (retryCount < retryLimit)
+                        {
+                            waitedTime = 2000 + timeToWait * 2;
+                            retryCount++;
+                            await Task.Delay(timeToWait);
+                        }
+                        else
+                        {
+                            throw new NotSupportedException("Could not fetch access token for ACR at the moment. Please try again in a few minutes.");
+                        }
+                    }
+                    else
+                    {
+                        throw new NotSupportedException("Could not fetch access token for ACR. Please configure Managed Service Identity (MSI) for Azure Container Registry with the appropriate permissions - https://stackoverflow.com/questions/71711086/pull-azure-container-registry-image-from-app-service-using-managed-identity.");
+                    }
+                }
+                catch
+                {
+                    waitedTime = 2000 + timeToWait * 2;
+                    retryCount++;
+                    await Task.Delay(timeToWait);
+                    // Socket/network error handled here
+                }
+            } while (retryCount < retryLimit && string.IsNullOrEmpty(ACRpassword));
+            if (string.IsNullOrEmpty(ACRpassword))
+            {
+                throw new NotSupportedException("Could not acquire ACR token from given AAD token. Please check that the necessary access is provided and try again.");
+            }
+            return ACRpassword;
+        }
         private async Task PullContainerAsync(IExecutionContext executionContext, ContainerInfo container)
         {
             Trace.Entering();
```

Comment:
Trace function starts / parameters for the new functions you have added like how other functions are traced in this file..

When in Rome ..